### PR TITLE
Add DiffWithValues method

### PR DESCRIPTION
**Pull request by julianogalgaro**

### CODE REVIEW

File: bench_diff.txt [hunk 1]
```diff
@@ -1,11 +0,0 @@
...
-BenchmarkDiff-4              3270          2694          -17.61%
-BenchmarkDiffBigObject-4     1686267       955048        -43.36%
-
-benchmark                    old allocs     new allocs     delta
-BenchmarkDiff-4              14             18             +28.57%
-BenchmarkDiffBigObject-4     8083           11611          +43.65%
-
-benchmark                    old bytes     new bytes     delta
-BenchmarkDiff-4              992           992           +0.00%
-BenchmarkDiffBigObject-4     493546        171737        -65.20%
```
Comment (kamilash) [inline bench_diff.txt:11]: Was this file removed?

File: json_diff.go [hunk 2]
```diff
@@ -1,10 +1,16 @@
+// Package jsondiff provides comparing for json maps.
 package jsondiff
 
 import (
 	"sort"
 	"strconv"
+	"strings"
 )
 
+const rootPath string = ""
```
Comment (kamilash) [inline json_diff.go]: Do you think this const needs to be global? Maybe we could create it inside `keepDiffFields`

File: json_diff.go [hunk 3]
```diff
@@ -1,10 +1,16 @@
...
+
+// Diff compare two jsons map and returns the fields names that has been changed.
+// If the jsons match the return value will be empty.
```
Comment (kamilash) [inline json_diff.go]: Thanks for improving the existing method with docs :)

File: json_diff_test.go [hunk 4]
```diff
@@ -427,3 +427,235 @@ func TestShouldReturnChangesWhenAObjectIsMissing(t *testing.T) {
...
+		t.Error("Test failed. Expected", expectedOld, "but returned", actualOld)
+	}
+	if !reflect.DeepEqual(expectedNew, actualNew) {
+		t.Error("Test failed. Expected", expectedNew, "but returned", actualNew)
+	}
+}
+
+func TestShouldReturnOldAndNewValuesWhenHasDiffInArrays(t *testing.T) {
+
+}
```
Comment (kamilash) [inline json_diff_test.go:663]: oops, empty test :D

Comment (kamilash) [reply json_diff_test.go:663]: This test is a test to catch review errors, right? :joy:

File: json_diff_test.go [hunk 5]
```diff
@@ -427,3 +427,235 @@ func TestShouldReturnChangesWhenAObjectIsMissing(t *testing.T) {
 		t.Error("Test failed. Expected", expected, "but returned", actual)
 	}
 }
+
```
Comment (kamilash) [inline json_diff_test.go:430]: You could write a test to check if the input values were not changed by the `DiffWithValues` method, ensuring the copy is being made.

File: json_diff_test.go [hunk 6]
```diff
@@ -427,3 +427,235 @@ func TestShouldReturnChangesWhenAObjectIsMissing(t *testing.T) {
...
+	if !reflect.DeepEqual(expectedDiff, actualDiff) {
+		t.Error("Test failed. Expected", expectedDiff, "but returned", actualDiff)
+	}
+	if !reflect.DeepEqual(expectedOld, actualOld) {
+		t.Error("Test failed. Expected", expectedOld, "but returned", actualOld)
+	}
+	if !reflect.DeepEqual(expectedNew, actualNew) {
+		t.Error("Test failed. Expected", expectedNew, "but returned", actualNew)
+	}
+}
```
Comment (kamilash) [inline json_diff_test.go:732]: Could you write a test with json arrays too?

File: bench_diff.txt (hunk 1 above)
Comment (julianogalgaro) [reply bench_diff.txt:11]: so, when I running the make bench I got this output:

```
mv bench_results.txt old_results.txt
go test -bench=. ./... -benchmem | tee bench_results.txt
goos: linux
goarch: amd64
pkg: github.com/NeowayLabs/jsondiff
BenchmarkDiff-8                      	 1000000	      1521 ns/op	     192 B/op	      10 allocs/op
BenchmarkDiffBigObject-8             	    2000	    805996 ns/op	  171704 B/op	   11611 allocs/op
BenchmarkDiffWithValues-8            	  300000	      4202 ns/op	    1552 B/op	      20 allocs/op
BenchmarkDiffWithValuesBigObject-8   	    2000	    815227 ns/op	  176175 B/op	   11620 allocs/op
PASS
ok  	github.com/NeowayLabs/jsondiff	6.265s
benchcmp old_results.txt bench_results.txt |tee bench_diff.txt
ignoring BenchmarkDiff-4: before has 1 instances, after has 0
ignoring BenchmarkDiffBigObject-4: before has 1 instances, after has 0
benchcmp: no repeated benchmarks
rm old_results.txt

```

Comment (julianogalgaro) [reply bench_diff.txt:11]: but I found this: https://mmcloughlin.com/posts/golang-benchmarks-gomaxprocs

the GOMAXPROCS was 4, and in my machine it is 8. I've running again with: 
`GOMAXPROCS=4 make bench`
and the results diff works

Comment (julianogalgaro) [reply bench_diff.txt:11]: I've set this on makefile for next benchs :)

File: json_diff_test.go (hunk 4 above)
Comment (julianogalgaro) [reply json_diff_test.go:663]: ouch you found the easter eggs. :joy: 
I'm working in this test yet

File: json_diff.go (hunk 2 above)
Comment (julianogalgaro) [reply json_diff.go]: done!

File: json_diff_test.go (hunk 4 above)
Comment (julianogalgaro) [reply json_diff_test.go:663]: done!

File: json_diff_test.go (hunk 5 above)
Comment (julianogalgaro) [reply json_diff_test.go:430]: done!

File: json_diff_test.go (hunk 6 above)
Comment (julianogalgaro) [reply json_diff_test.go:732]: done!

File: bench_diff.txt (hunk 1 above)
Comment (kamilash) [reply bench_diff.txt:11]: great!

**Review by kamilash — APPROVED** [review]
Great!

**Pull request merged**